src/raft_log.rs

```diff
@@ -503,6 +503,8 @@ impl<T: Storage> RaftLog<T> {
             snapshot_index = snapshot.get_metadata().index,
             snapshot_term = snapshot.get_metadata().term,
         );
+        // NOTE: here we don't change applied index because some applications apply snapshot
+        // asynchronous.
```

Comment:
It's not about asynchronous, it's just about correctness. Apply index should be updated after data is restored, no matter if it's done asynchronously no not.